crates/bevy_pbr/src/render/mesh_bindings.rs

```diff
@@ -1,229 +1,351 @@
+            multisampled: false,
+        },
+        count: None,
+    }
+}
+fn targets_entry(binding: u32, texture: &TextureView) -> BindGroupEntry {
+    BindGroupEntry {
+        binding,
+        resource: BindingResource::TextureView(texture),
+    }
+}
+
+/// Create a layout for a specific [`ActiveVariant`] by combining the layouts of
+/// all the active shader features.
+fn variant_layout(active_variant: ActiveVariant, device: &RenderDevice) -> BindGroupLayout {
+    let skin = active_variant.contains(ActiveVariant::SKIN);
+    let morph = active_variant.contains(ActiveVariant::MORPH);
+    let motion_vectors = active_variant.contains(ActiveVariant::MOTION_VECTORS);
+
+    let mut layout_entries = SmallVec::<[BindGroupLayoutEntry; 6]>::new();
+
+    layout_entries.push(model_layout(device, 0));
+
+    if skin {
+        layout_entries.push(skinning_layout(1));
     }
-    pub(super) fn model(binding: u32, resource: BindingResource) -> BindGroupEntry {
-        BindGroupEntry { binding, resource }
+    if morph {
+        layout_entries.extend([weights_layout(2), targets_layout(3)]);
     }
-    pub(super) fn skinning(binding: u32, buffer: &Buffer) -> BindGroupEntry {
-        entry(binding, JOINT_BUFFER_SIZE as u64, buffer)
+    if skin && motion_vectors {
+        layout_entries.push(skinning_layout(4));
     }
-    pub(super) fn weights(binding: u32, buffer: &Buffer) -> BindGroupEntry {
-        entry(binding, MORPH_BUFFER_SIZE as u64, buffer)
+    if morph && motion_vectors {
+        layout_entries.push(weights_layout(5));
     }
-    pub(super) fn targets(binding: u32, texture: &TextureView) -> BindGroupEntry {
-        BindGroupEntry {
-            binding,
-            resource: BindingResource::TextureView(texture),
+
+    device.create_bind_group_layout(&BindGroupLayoutDescriptor {
+        entries: &layout_entries,
+        label: Some(&*format!("{active_variant}_layout")),
+    })
+}
+
+/// Create [`BindGroup`]s and add them to a [`MeshBindGroups`].
+///
+/// Use [`MeshBindGroups::bind_group_builder`] to get a [`MeshBindGroupBuilder`],
+/// and use [`MeshBindGroupBuilder::add_variant`] to add new bind group variants.
+pub struct MeshBindGroupBuilder<'a> {
+    layouts: &'a MeshLayouts,
+    device: &'a RenderDevice,
+    model: BindingResource<'a>,
+    pub bind_groups: &'a mut MeshBindGroups,
+}
+impl<'a> MeshBindGroupBuilder<'a> {
+    /// Create a [`BindGroup`] with the provided active features, and add it to [`Self::bind_groups`].
+    ///
+    /// Each parameter to `add_variant` is a shader feature. When the parameter
+    /// is `None`, it means that the shader feature is disabled.
+    pub fn add_variant(
+        &mut self,
+        skin: Option<&Buffer>,
+        morph: Option<(AssetId<Mesh>, &Buffer, &TextureView)>,
+        previous_skin: Option<&Buffer>,
+        previous_weights: Option<&Buffer>,
+    ) {
+        let mut bind_group_entries = SmallVec::<[BindGroupEntry; 6]>::new();
```

Comment:
Not sure what benefit SmallVec brings here other than overhead / an unnecessary dependency as this code would not exceed the array on the stack. I'd just use an array and an index variable.

Reply:
I would have personally used `ArrayVec` from the `arrayvec` crate, but `SmallVec` is already a dependency of `bevy_pbr`, so why not use it?

The advantage compared to using an array and subslicing it is that we don't have to initialize zero-valued `BindGroupEntry` (`BindGroupEntry` does not implement `Default`, so it would require a lot of boilerplate code). We could use `[MaybeUninit<BindGroupEntry>; 6]`, but it would also require a bit of boilerplate and `unsafe` unrelated to what we want to accomplish, and that's pretty much what `SmallVec` already does for us regardless.